openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,435 @@
+			managedResource := fmt.Sprintf("%s.%scores", workloadType, podWorkloadAnnotationPrefix)
+			newCPURequest := resource.NewMilliQuantity(cpuRequestInMilli*1000, cpuRequest.Format)
+			c.Resources.Requests[coreapi.ResourceName(managedResource)] = *newCPURequest
+			c.Resources.Limits[coreapi.ResourceName(managedResource)] = *newCPURequest
+		}
+	}
+}
+
+func isGuaranteed(containers []coreapi.Container) bool {
+	for _, c := range containers {
+		// only memory and CPU resources are relevant to decide pod QoS class
+		for _, r := range []coreapi.ResourceName{coreapi.ResourceMemory, coreapi.ResourceCPU} {
+			limit := c.Resources.Limits[r]
+			request, requestExist := c.Resources.Requests[r]
+
+			if limit.IsZero() {
+				return false
+			}
+
+			if !requestExist {
+				continue
+			}
+
+			// it some corner case, when you set CPU request to 0 the k8s will change it to the value
+			// specified under the limit
+			if r == coreapi.ResourceCPU && request.IsZero() {
+				continue
+			}
+
+			if !limit.Equal(request) {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func isBestEffort(containers []coreapi.Container) bool {
+	for _, c := range containers {
+		// only memory and CPU resources are relevant to decide pod QoS class
+		for _, r := range []coreapi.ResourceName{coreapi.ResourceMemory, coreapi.ResourceCPU} {
+			limit := c.Resources.Limits[r]
+			request := c.Resources.Requests[r]
+
+			if !limit.IsZero() || !request.IsZero() {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func getPodQoSClass(containers []coreapi.Container) coreapi.PodQOSClass {
+	if isGuaranteed(containers) {
+		return coreapi.PodQOSGuaranteed
+	}
+
+	if isBestEffort(containers) {
+		return coreapi.PodQOSBestEffort
+	}
+
+	return coreapi.PodQOSBurstable
+}
+
+func isManagementResourceAvailableForAllNodes(nodes []*corev1.Node, workloadType string) error {
+	managedResource := fmt.Sprintf("%s.%scores", workloadType, podWorkloadAnnotationPrefix)
+	for _, node := range nodes {
+		if v, ok := node.Status.Allocatable[corev1.ResourceName(managedResource)]; !ok || v.IsZero() {
+			return fmt.Errorf(admissionWarningNoManagedResource, node.Name, managedResource)
+		}
+	}
+
+	return nil
+}
+
+func isNamespaceAllowed(annotations map[string]string, workloadType string) bool {
+	v, found := annotations[namespaceAllowedAnnotation]
+	if !found {
+		return false
+	}
+
+	for _, t := range strings.Split(v, ",") {
+		if workloadType == t {
+			return true
+		}
+	}
+
+	return false
+}
+
+func getWorkloadAnnotations(annotations map[string]string) map[string]string {
+	workloadAnnotations := map[string]string{}
+	for k, v := range annotations {
+		if strings.HasPrefix(k, podWorkloadAnnotationPrefix) {
+			workloadAnnotations[k] = v
+		}
+	}
+
+	return workloadAnnotations
+}
+
+func getWorkloadType(workloadAnnotations map[string]string) (string, error) {
+	for k := range workloadAnnotations {
```

Comment:
I expect this loop to return an error if more than one workloadtype is specified.

Reply:
In the future it may not be an error to have multiple annotations and this function would be expected to choose the right one to use. We could centralize all of the logic related to that here, though, by moving the length test into this function so that when we do support multiple annotations in the future we only have to change one part of the code.

What do you think @cynepco3hahue ?

Reply:
done, moved the relevant logic under `getWorkloadType`.